### ISSUE TITLE
jaxlib==0.1.61 on CPU/GPU computes incorrect values with jax.lax.cond inside jax.lax.scan

### ISSUE THREAD

**Issue by mblondel** [google/jax#5900]
In the following code, `jax.lax.scan` gives an incorrect gradient while `scan` copy-pasted from the documentation gives the correct gradient (tested against finite difference). The correct gradient is `108` but jax returns `324.0`.

```python

import jax
import jax.numpy as jnp


def scan(f, init, xs, length=None):
  if xs is None:
    xs = [None] * length
  carry = init
  ys = []
  for x in xs:
    carry, y = f(carry, x)
    ys.append(y)
  return carry, None


def _while_loop(cond_fun, body_fun, init_val, max_iter, scan_fun):
  def _iter(val):
    next_val = body_fun(val)
    next_cond = cond_fun(next_val)
    return next_val, next_cond

  def _fun(tup, it):
    val, cond = tup
    # When cond is met, we start doing noops.
    return jax.lax.cond(cond, _iter, lambda x: (x, False), val), it

  init = (init_val, cond_fun(init_val))
  return scan_fun(_fun, init, None, length=max_iter)[0][0]


def test1(scan_fun):
  def my_pow(x, y):
    def body_fun(val):
      return val * x
    def cond_fun(val):
      return True

    return _while_loop(cond_fun=cond_fun, body_fun=body_fun, init_val=1.0,
                       max_iter=y, scan_fun=scan_fun)

  print(my_pow(3, 4), pow(3, 4))

  # Correct gradient when using scan_fun = scan
  # Incorrect gradient when using scan_fun = jax.lax.scan
  eps = 1e-5
  print(jax.grad(my_pow)(3.0, 4),
        (my_pow(3.0 + eps, 4) - my_pow(3.0 - eps, 4)) / (2 * eps))


def test2(scan_fun):
  def my_pow(x, y, max_val):
    def body_fun(val):
      return val * x
    def cond_fun(val):
      return val < max_val

    return _while_loop(cond_fun=cond_fun, body_fun=body_fun, init_val=1.0,
                      max_iter=y, scan_fun=scan_fun)

  # We asked for pow(3, 6) but due to max_val, we get pow(3, 4).
  print(my_pow(3, 6, max_val=81), pow(3, 4))

  # Correct gradient when using scan_fun = scan
  # Incorrect gradient when using scan_fun = jax.lax.scan
  print(jax.grad(my_pow)(3.0, 6, max_val=81),
        jax.grad(jnp.power)(3.0, 4))


if __name__ == '__main__':
  scan_fun = jax.lax.scan
  #scan_fun = scan

  print("test1")
  test1(scan_fun)

  print("test2")
  test2(scan_fun)
```

**Comment by mattjj** [comment]
Thanks for the report!

The same doesn't seem to happen with forward-mode: I changed the calls to` jax.grad` to calls to `deriv:

```python
def deriv(f):
  return lambda x, *args,  **kwargs: jax.jvp(lambda x: f(x, *args, **kwargs), (x,), (1.0,))[1]
```

And got this:

```
test1
81.0 81
108.0 108.3374
test2
81.0 81
108.0 108.0
```

That suggests to me there's an issue with transposition here...

**Comment by mattjj** [comment]
Wow, this had us scratching our heads for a bit, since everything looked fine on the JAX side. Then we realized: this might be an XLA:CPU bug! Two strong lines of evidence to suggest that:
1. I ran it on a TPU colab and there was no discrepancy;
2. if we change [this line](https://github.com/google/jax/blob/249cd8b7f77630597b4c0e8a64f40bafc2d9f12e/jax/_src/lax/control_flow.py#L1379) to call `_scan_impl_unrolled`, which is a pure Python evaluation procedure for `scan`,  instead of `_scan_impl_loop`, which executes an XLA While loop, then there's no discrepancy.

So I think we've got to file this with XLA:CPU folks, as there may be a miscompile of this loop-of-cond program on CPU.

I slightly simplified the repro to this:

```python
import jax
import jax.numpy as jnp


def deriv(f):
  return lambda x, *args: jax.jvp(lambda x: f(x, *args), (x,), (1.0,))[1]

def deriv2(f):
  return lambda x, *args: jax.linearize(lambda x: f(x, *args), x)[1](1.0)

def _while_loop(cond_fun, body_fun, init_val, max_iter):
  def _iter(val):
    return body_fun(val), cond_fun(val)

  def _fun(tup, _):
    val, cond = tup
    # jax.lax.cond(cond, _iter, _iter, val), _   # this works!
    return jax.lax.cond(cond, _iter, lambda x: (x, False), val), _

  init = (init_val, cond_fun(init_val))
  return jax.lax.scan(_fun, init, None, length=max_iter)[0][0]


def test1():
  def my_pow(x, y):
    def body_fun(val):
      return val * x
    def cond_fun(val):
      return True

    return _while_loop(cond_fun, body_fun, 1.0, y)

  x = 3.0
  n = 1

  eps = 1e-5
  print(deriv(my_pow)(x, n),
        (my_pow(x + eps, n) - my_pow(x - eps, n)) / (2 * eps))
  print(deriv2(my_pow)(x, n),
        (my_pow(x + eps, n) - my_pow(x - eps, n)) / (2 * eps))


test1()
```

```
1.0 1.001358
3.0 1.001358
```

Notice that `jvp` works, while `linearize` doesn't. That is, we don't even need to involve reverse-mode (and transposition). 

As in the OP here, the predicate value to the conditional is always `True`.

**Comment by mattjj** [comment]
Ah, another twist to the story! This is definitely an XLA bug, but it's already fixed at HEAD. That is, if you build jaxlib yourself this bug won't occur. The problem is it's not fixed in our jaxlib== binary release.

For Google-internal folks: the test in cl/356908804, basically with the above repro but set up to easily test on CPU/GPU/TPU, passes at HEAD. But if we sync back to cl/361191651, which corresponds to the latest jaxlib==0.1.61 (found by looking up the commit in this repo corresponding to that release, checking WORKSPACE for the TF git commit hash, then looking up the TF commit message in that repository), then we get an error, though only on CPU and not GPU or TPU.

So the fix for this is for us to release a new jaxlib. The workaround is to build your own jaxlib. (Unfortunately, downgrading doesn't work, as this bug seems to be present in jaxlib==0.1.60, and jaxlib==0.1.59 isn't compatible with the master version of jax so I didn't even check it.)

**Comment by mattjj** [comment]
I forgot to mention: big thanks to @hawkinsp for guiding me through this debugging process. I learned a ton.

**Comment by mattjj** [comment]
I set #5956 to close this issue because we're confident that the next jaxlib release will fix it (as will building jaxlib from master right now). We'll work on a jaxlib release now!

**Comment by mblondel** [comment]
Thanks for the investigation and for the non-regression test!
